File: examples/draw-triangle.cpp
#include "AppBase.hpp"
#include "ren/baking/mesh.hpp"

#include <glm/glm.hpp>
#include <utility>

class DrawTriangleApp : public AppBase {
  ren::Handle<ren::Mesh> m_mesh;
  ren::Handle<ren::Material> m_material;
  ren::Handle<ren::MeshInstance> m_triangle;

public:
  void init() {
    AppBase::init("Draw Triangle");

    ren::Scene *scene = get_scene();

    glm::vec3 positions[] = {
        {0.0f, 0.5f, 0.0f},
        {-std::sqrt(3.0f) / 4.0f, -0.25f, 0.0f},
        {std::sqrt(3.0f) / 4.0f, -0.25f, 0.0f},
    };

    glm::vec4 colors[] = {
        {1.0f, 0.0f, 0.0f, 1.0f},
        {0.0f, 1.0f, 0.0f, 1.0f},
        {0.0f, 0.0f, 1.0f, 1.0f},
    };

    glm::vec3 normals[] = {
        {0.0f, 0.0f, 1.0f},
        {0.0f, 0.0f, 1.0f},
        {0.0f, 0.0f, 1.0f},
    };

    ren::ScratchArena scratch;
    ren::Blob blob =
        ren::bake_mesh_to_memory(scratch, {
                                              .num_vertices = 3,
                                              .positions = positions,
                                              .normals = normals,
                                              .colors = colors,
                                          });
    m_mesh = create_mesh(&m_frame_arena, scene, blob.data, blob.size);

    m_material = ren::create_material(&m_frame_arena, scene,
                                      {
                                          .roughness_factor = 0.5f,
                                          .metallic_factor = 1.0f,
                                      });

    // Ambient day light
    std::ignore =
        ren::create_directional_light(scene, {
                                                 .color = {1.0f, 1.0f, 1.0f},
                                                 .illuminance = 25'000.0f,
                                                 .origin = {0.0f, 0.0f, 1.0f},
                                             });

    ren::Handle<ren::Camera> camera = get_camera();

    ren::set_camera_orthographic_projection(scene, camera, {.width = 2.0f});
    ren::set_camera_transform(scene, camera,
                              {
                                  .position = {0.0f, 0.0f, 1.0f},
                                  .forward = {0.0f, 0.0f, -1.0f},
                                  .up = {0.0f, 1.0f, 0.0f},
                              });
  }

  void process_frame(ren::u64 dt_ns) override {
    ren::Scene *scene = get_scene();
    ren::destroy_mesh_instance(&m_frame_arena, scene, m_triangle);
    m_triangle =
        ren::create_mesh_instance(&m_frame_arena, scene, {m_mesh, m_material});
    ren::set_mesh_instance_transform(&m_frame_arena, scene, m_triangle,
                                     glm::mat4(1.0f));
  }

  static void run() { AppBase::run<DrawTriangleApp>(); }
};

int main() {
  ren::ScratchArena::init_for_thread();
  DrawTriangleApp::run();
}

File: rust/examples/draw-triangle.rs
use anyhow::Result;
use ren::{
    CameraDesc, CameraProjection, DirLightDesc, MaterialDesc, MeshDesc, MeshInstDesc, Scene,
};
use utils::{App, AppBase};

mod utils;

struct DrawTriangleApp {}

impl DrawTriangleApp {
    fn new(scene: &mut Scene) -> Result<Self> {
        let positions = [
            [0.0, 0.5, 0.0],
            [-(3.0f32).sqrt() / 4.0, -0.25, 0.0],
            [3.0f32.sqrt() / 4.0, -0.25, 0.0],
        ];

        let colors = [
            [1.0, 0.0, 0.0, 1.0],
            [0.0, 1.0, 0.0, 1.0],
            [0.0, 0.0, 1.0, 1.0],
        ];

        let normals = [[0.0, 0.0, 1.0], [0.0, 0.0, 1.0], [0.0, 0.0, 1.0]];

        let indices = [0, 1, 2];

        let mesh = scene.create_mesh(&MeshDesc {
            positions: &positions,
            colors: Some(&colors),
            normals: Some(&normals),
            indices: Some(&indices),
            ..Default::default()
        })?;

        let material = scene.create_material(&MaterialDesc {
            metallic_factor: 1.0,
            roughness_factor: 0.5,
            ..Default::default()
        })?;

        scene.create_mesh_inst(&MeshInstDesc {
            mesh,
            material,
            casts_shadows: false,
        })?;

        scene.create_dir_light(&DirLightDesc {
            color: [1.0, 1.0, 1.0],
            illuminance: 25_000.0,
            origin: [0.0, 0.0, 1.0],
        })?;

        Ok(Self {})
    }
}

impl App for DrawTriangleApp {
    fn get_title(&self) -> &str {
        "Draw Triangle"
    }

    fn handle_frame(&mut self, scene: &mut Scene, width: u32, height: u32) -> Result<()> {
        scene.set_camera(&CameraDesc {
            projection: CameraProjection::Orthographic { width: 2.0 },
            width,
            height,
            position: [0.0, 0.0, 1.0],
            forward: [0.0, 0.0, -1.0],
            up: [0.0, 1.0, 0.0],
            exposure_compensation: 3.0,
            ..Default::default()
        })?;
        Ok(())
    }
}

fn main() -> Result<()> {
    let mut base = AppBase::new()?;
    let app = DrawTriangleApp::new(base.get_scene_mut())?;
    base.run(app);
}
